pkg/fs_algo/fs_algo/fs_perf_viz.py

```diff
@@ -0,0 +1,428 @@
+
+    # Loop through all datasets
+    for ds in datasets:
+        path_meta_pred = f'{path_meta_pred}'.format(ds = ds, dir_std_base = dir_std_base, ds_type = ds_type, write_type = write_type)
+        meta_pred = pd.read_parquet(path_meta_pred)
+
+        # Loop through all algorithms
+        for algo in algos:
+            # Loop through all metrics
+            for metric in metrics:
+                # Pull the predictions
+                path_pred = fsate.std_pred_path(dir_out,algo=algo,metric=metric,dataset_id=ds)
+                pred = pd.read_parquet(path_pred)
+                data = pd.merge(meta_pred, pred, how = 'inner', on = 'comid')
+                os.makedirs(f'{dir_out}/data_visualizations', exist_ok= True)
+                # If you want to export the merged data for any reason: 
+                # data.to_csv(f'{dir_out}/data_visualizations/{ds}_{algo}_{metric}_data.csv')
+
+                # Does the user want a scatter plot comparing the observed module performance and the predicted module performance by RaFTS?
+                if 'perf_map' in true_keys:
+                    states = gpd.read_file('/Users/laurenbolotin/data/conus_states_census.shp')
+                    states = states.to_crs("EPSG:4326")
+
+                    # Plot performance on map
+                    lat = data['Y']
+                    lon = data['X']
+                    geometry = [Point(xy) for xy in zip(lon,lat)]
+                    geo_df = gpd.GeoDataFrame(geometry = geometry)
+                    geo_df['performance'] = data['prediction'].values
+                    geo_df.crs = ("EPSG:4326")
+
+                    fig, ax = plt.subplots(1, 1, figsize=(20, 24))
+                    base = states.boundary.plot(ax=ax,color="#555555", linewidth=1)
+                    # Points
+                    geo_df.plot(column="performance", ax=ax, markersize=150, cmap='viridis', legend=False, zorder=2) # delete zorder to plot points behind states boundaries
+                    # States
+                    states.boundary.plot(ax=ax, color="#555555", linewidth=1, zorder=1)  # Plot states boundary again with lower zorder
+
+                    cbar = plt.cm.ScalarMappable(norm=matplotlib.colors.Normalize(vmin=-0.41,vmax = 1), cmap='viridis')
+                    ax.tick_params(axis='x', labelsize= 24)
+                    ax.tick_params(axis='y', labelsize= 24)
+                    plt.xlabel('Latitude',fontsize = 26)
+                    plt.ylabel('Longitude',fontsize = 26)
+                    cbar_ax = plt.colorbar(cbar, ax=ax,fraction=0.02, pad=0.04)
+                    cbar_ax.set_label(label=metric,size=24)
+                    cbar_ax.ax.tick_params(labelsize=24)  # Set colorbar tick labels size
+                    plt.title("Predicted Performance: {}".format(ds), fontsize = 28)
+
+                    # Save the plot as a .png file
+                    output_path = f'{dir_out}/data_visualizations/{ds}_{algo}_{metric}_performance_map.png'
```

Comment:
Return a matplot lib figure as an object (plt.gcf())
Create functions that define paths
Create functions that do the file I/O so that when we migrate to AWS we can just update the function
Move the creation of the data_visualizations folder into a function that creates the path for it
dir_ * if a directory
path_* if a path
Could include smaller functions for creating the path to the performance map, for example: fsate lines 492, 474 
Look at plot learning curve for example: https://github.com/glitt13/formulation-selector/blob/agu_24/pkg/fs_algo/fs_algo/fs_algo_train_eval.py

Reply:
Classes are mostly for conveniently passing along data, so you can probably just do functions for now and either make a separate plotting file or add to fsate